stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -697,6 +697,8 @@ end
 *(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) = _mapreduce_prod(*, x, D, y)
 dot(x::AbstractVector, D::Diagonal, y::AbstractVector) = _mapreduce_prod(dot, x, D, y)
 
+dot(A::Diagonal, B::Diagonal) = dot(A.diag, B.diag)
```

Comment:
```suggestion
dot(A::Diagonal, B::Diagonal) = dot(A.diag, B.diag)
dot(A::Diagonal, B::AbstractMatrix) = dot(A, Diagonal(B))
dot(A::AbstractMatrix, B::Diagonal) = conj(dot(B, A))
```
Not sure, but this would be an easy approach. It allocates the `diag` vector, though.

Reply:
I agree with the approach, but I don't think `dot` methods should allocate.
`Diagonal(view(B, diagind(B)))` seems to work

Reply:
(technically it allocates a constant amount of 80 bytes):
```julia
julia> M = rand(30_000,30_000);

julia> @btime Diagonal(view($M, diagind($M)));
  49.629 ns (2 allocations: 80 bytes)
```

Reply:
Ok, this requires further size checks... This could be an alternative, that only allocates the diagonal indices (a range):
```julia
function dot(D::Diagonal, B::AbstractMatrix)
    length(D) == length(B) || throw(DimensionMismatch(...))
    return dot(D.diag, view(B, diagind(B)))
end
```

Reply:
That's right, it allocates the range object, independent of the size of the matrix.

Reply:
Well, or the view object, but anyway.